File: output/gemma4-e2b/avatar/Java/Python/atcoder_ABC165_C.py
import sys

def solve():
    # Read all input from stdin
    data = sys.stdin.read().split()
    if not data:
        return

    # Initialize pointers for reading data
    data_ptr = 0

    # Read N, M, Q
    N = int(data[data_ptr])
    M = int(data[data_ptr + 1])
    Q = int(data[data_ptr + 2])
    data_ptr += 3

    # Initialize arrays a, b, c, d
    A = [0] * Q
    B = [0] * Q
    C = [0] * Q
    D = [0] * Q

    # Read constraints
    for i in range(Q):
        A[i] = int(data[data_ptr]) - 1
        B[i] = int(data[data_ptr + 1]) - 1
        C[i] = int(data[data_ptr + 2])
        D[i] = int(data[data_ptr + 3])
        data_ptr += 4

    # Global variable to store the maximum answer (equivalent to static long ans)
    max_ans = -100L

    def dfs(current_list):
        nonlocal max_ans
        
        if len(current_list) == N:
            score = 0L
            for i in range(Q):
                # Check the condition: list[B[i]] - list[A[i]] == C[i]
                if current_list[B[i]] - current_list[A[i]] == C[i]:
                    score += D[i]
            
            if score > max_ans:
                max_ans = score
            return

        # Recursive step: Try adding numbers from 1 to M
        
        # Determine the starting number for iteration
        if not current_list:
            start_num = 1
        else:
            # Java logic: starts from the last added element + 1
            start_num = current_list[-1] + 1
        
        for num in range(start_num, M + 1):
            # Add num to the list
            current_list.append(num)
            
            # Recurse
            dfs(current_list)
            
            # Backtrack: Remove num
            current_list.pop()

    # Start the search
    dfs([])

    print(max_ans)

solve()

File: output/gemma4-e2b/avatar/Python/Java/atcoder_ABC165_C.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.List;

public class atcoder_ABC165_C {

    // Global variables to store input parameters
    static int N, M, Q;
    static int[] a, b, c, d;
    static long res = 0; // Use long for result just in case sums get large, although int might suffice based on Python context.

    /**
     * Recursive function to explore sequences A.
     * A is the current sequence being built.
     */
    private static void rec(List<Integer> A) {
        if (A.size() == N) {
            long ans = 0;
            for (int i = 0; i < Q; i++) {
                // Python: A[b[i] - 1] - A[a[i] - 1] == c[i]
                // Java uses 0-based indexing. Indices a[i] and b[i] are 1-based in the problem context.
                if (A.get(b[i] - 1) - A.get(a[i] - 1) == c[i]) {
                    ans += d[i];
                }
            }
            if (ans > res) {
                res = ans;
            }
            return;
        }

        // Determine the starting value for the next element
        int lastMax = (A.isEmpty()) ? 1 : A.get(A.size() - 1);

        // Iterate possible values for the next element (i) from last_max to M
        for (int i = lastMax; i <= M; i++) {
            // Create a new list for the recursive call (A + [i])
            List<Integer> nextA = new ArrayList<>(A);
            nextA.add(i);
            rec(nextA);
        }
    }

    public static void solve() {
        // Start the recursion with an empty list
        rec(new ArrayList<>());
        System.out.println(res);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line;

        // Read all tokens from stdin
        StringTokenizer st;

        // Read N, M, Q
        if ((st = new StringTokenizer(br.readLine())) == null) return;
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        Q = Integer.parseInt(st.nextToken());

        // Initialize arrays a, b, c, d
        a = new int[Q];
        b = new int[Q];
        c = new int[Q];
        d = new int[Q];

        // Read the Q sets of a, b, c, d
        for (int i = 0; i < Q; i++) {
            if ((st = new StringTokenizer(br.readLine())) == null) return;
            a[i] = Integer.parseInt(st.nextToken());
            b[i] = Integer.parseInt(st.nextToken());
            c[i] = Integer.parseInt(st.nextToken());
            d[i] = Integer.parseInt